apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,417 @@
+	if err != nil {
+		return nil, nil, errors.Errorf("validating resource size and hash: %w", err)
+	}
+
+	return reader, &uploadedResource{
+		UUID:        res.UUID,
+		Application: uReq.Application,
+		Resource:    res.Resource,
+		Size:        uReq.Size,
+		Fingerprint: uReq.Fingerprint,
+	}, nil
+}
+
+// extractUploadRequest pulls the required info from the HTTP request.
+func extractUploadRequest(req *http.Request) (api.UploadRequest, error) {
+	var ur api.UploadRequest
+
+	if req.Header.Get(api.HeaderContentLength) == "" {
+		size := req.ContentLength
+		// size will be negative if there is no content.
+		if size < 0 {
+			size = 0
+		}
+		req.Header.Set(api.HeaderContentLength, fmt.Sprint(size))
+	}
+
+	ctype := req.Header.Get(api.HeaderContentType)
+	if ctype != api.ContentTypeRaw {
+		return ur, errors.Errorf("unsupported content type %q", ctype)
+	}
+
+	application, name := extractEndpointDetails(req.URL)
+	fingerprint := req.Header.Get(api.HeaderContentSha384) // This parallels "Content-MD5".
+	sizeRaw := req.Header.Get(api.HeaderContentLength)
+
+	fp, err := charmresource.ParseFingerprint(fingerprint)
+	if err != nil {
+		return ur, errors.Errorf("parsing fingerprint: %w", err)
+	}
+
+	filename, err := extractFilename(req)
+	if err != nil {
+		return ur, errors.Capture(err)
+	}
+
+	size, err := strconv.ParseInt(sizeRaw, 10, 64)
+	if err != nil {
+		return ur, errors.Errorf("parsing size: %w", err)
+	}
+
+	ur = api.UploadRequest{
+		Application: application,
+		Name:        name,
+		Filename:    filename,
+		Size:        size,
+		Fingerprint: fp,
+	}
+	return ur, nil
+}
+
+// extractEndpointDetails pulls the endpoint wildcard values from
+// the provided URL.
+func extractEndpointDetails(url *url.URL) (application, name string) {
+	application = url.Query().Get(":application")
+	name = url.Query().Get(":resource")
+	return application, name
+}
+
+func extractFilename(req *http.Request) (string, error) {
+	disp := req.Header.Get(api.HeaderContentDisposition)
+
+	// the first value returned here is the media type Name (e.g. "form-data"),
+	// but we don't really care.
+	_, vals, err := mime.ParseMediaType(disp)
+	if err != nil {
+		return "", errors.Errorf("badly formatted Content-Disposition: %w", err)
+	}
+
+	param, ok := vals[api.FilenameParamForContentDispositionHeader]
+	if !ok {
+		return "", errors.Errorf("missing filename in resource upload request")
+	}
+
+	filename, err := decodeParam(param)
+	if err != nil {
+		return "", errors.Errorf("decoding filename %q from upload request: %w", param, err)
+	}
+	return filename, nil
+}
+
+func decodeParam(s string) (string, error) {
+	decoded, err := new(mime.WordDecoder).Decode(s)
+
+	// If encoding is not required, the encoder will return the original string.
+	// However, the decoder doesn't expect that, so it barfs on non-encoded
+	// strings. To detect if a string was not encoded, we simply try encoding
+	// again, if it returns the same string, we know it wasn't encoded.
```

Comment:
What could we be encoding here? What is the param? This feels like someone should be either base64 the result, or validating the parameter.